rfcs/syntax-string-interpolation.md

```diff
@@ -0,0 +1,140 @@
+4. No support for values implementing the `__tostring` metamethod.
+5. Using `%` is in itself a dangerous operation within `string.format`.
+   * `"Your health is %d% so you need to heal up."` causes a runtime error because `% so` is actually parsed as `(%s)o` and now requires a corresponding string.
+6. Having to use parentheses around string literals just to call a method of it.
+
+## Design
+
+To fix all of those issues, we need to do a few things.
+
+1. A new string interpolation expression (fixes #5, #6)
+2. Extend `string.format` to accept values of arbitrary types (fixes #1, #2, #3, #4)
+
+Because we care about backward compatibility, we need some new syntax in order to not change the meaning of existing strings. There are a few components of this new expression:
+
+1. A string chunk (`` `...{ ``, `}...{`, and `` }...` ``) where `...` is a range of 0 to many characters.
+   * `%` escapes `%` and `{`, whereas `\` escapes `` ` ``.
+   * Restriction: the string interpolation literal must have at least one value to interpolate. We do not need 3 ways to express a single line string literal.
+   * The pairs must be on the same line (unless a `\` escapes the newline) but expressions needn't be on the same line.
+2. An expression between the braces. This is the value that will be interpolated into the string.
+3. Formatting specification may follow the expression, delimited by an unambiguous character.
+   * Restriction: the formatting specification must be constant at parse time.
+   * The syntax for formatting specification and a delimiter is undefined in this RFC. A future RFC will implement this.
+
+To put the above into formal EBNF grammar:
+
+```
+stringinterp ::= <INTERP_BEGIN> exp {<INTERP_MID> exp} <INTERP_END>
+```
+
+Which, in actual Luau code, will look like the following:
+
+```
+local world = "world"
+print(`Hello {world}!`)
+--> Hello world!
+
+local combo = {5, 2, 8, 9}
+print(`The lock combinations are: {table.concat(combo, ", ")}`)
+--> The lock combinations are: 5, 2, 8, 9
+
+local set1 = Set.new({0, 1, 3})
+local set2 = Set.new({0, 5, 4})
+print(`{set1} ∪ {set2} = {Set.union(set1, set2)}`)
+--> {0, 1, 3} ∪ {0, 5, 4} = {0, 1, 3, 4, 5}
+
+print(`Some example escaping the braces %{like so}`)
+print(`% that doesn't escape anything is part of the string...`)
+print(`%% will escape the second %...`)
+print(`Some text that also includes \`...`)
+--> Some example escaping the braces {like so}
+--> % that doesn't escape anything is part of the string...
+--> % will escape the second %...
+--> Some text that also includes `...
+```
+
+As for how newlines are handled, they are handled the same as other string literals. Any text between the `{}` delimiters are not considered part of the string, hence newlines are OK. The main thing is that one opening pair will scan until either a closing pair is encountered, or an unescaped newline.
+
+```
+local name = "Luau"
+
+print(`Welcome to {
+    Luau
```

Comment:
Shouldn't it be `name`?

Same for 84. and 93. line.

Reply:
Yeah. It should be. Good catch.

Reply:
Fixed.